### PR TITLE
Added optional leading backslash to YAML frontmatter regex pattern

### PR DESCRIPTION
**Pull request by DarylWright**
When using YAML frontmatter with Haml files, the string '---' is considered invalid syntax. The hyphen is reserved for executing Ruby code and needs to be escaped to be output as plain text ('\\---').

### CODE REVIEW

**Comment by Arcovion** [comment]
Frontmatter should already be removed before the file is passed to template engines, this must be a bug.

File: middleman-core/lib/middleman-core/util/data.rb [hunk 1]
```diff
@@ -57,6 +57,7 @@ def parse(full_path, known_type=nil)
   # @return [Hash]
   Contract String, Pathname, Bool => Hash
   def parse_yaml(content, full_path)
+    content.sub!(/\\---/, '---')
```
Comment (Arcovion) [inline middleman-core/lib/middleman-core/util/data.rb]: `parse_yaml` should only be called on valid YAML/JSON files (backslash makes it not valid YAML), so this shouldn't be here... are you calling `parse_yaml` directly from an extension or somewhere else?

**Comment by DarylWright** [comment]
Thanks for jumping in! The frontmatter gets removed no problem, that isn't the issue here. What is the issue is that the usage of YAML frontmatter in Haml forces you to save the file with an incorrect syntax and also shows up as errors in editors such as Sublime Text and RubyMine. The hyphen character is reserved in Haml for executing Ruby code and needs to be escaped to be interpreted as plain text.

File: middleman-core/lib/middleman-core/util/data.rb (hunk 1 above)
Comment (DarylWright) [inline middleman-core/lib/middleman-core/util/data.rb]: Yes, I see my error now, thanks. I see now how I can modify the frontmatter in the parse method.

File: middleman-core/lib/middleman-core/util/data.rb [hunk 2]
```diff
@@ -43,7 +43,7 @@ def parse(full_path, known_type=nil)
     end
 
     case [start, stop]
-    when %w(--- ---), %w(--- ...)
+    when %w(--- ---), %w(--- ...), %(\\--- \\---), %w(\\--- ...)
       [parse_yaml(frontmatter, full_path), additional_content]
```
Comment (DarylWright) [inline middleman-core/lib/middleman-core/util/data.rb]: I'd imagine that I'd have to strip the backslash from the frontmatter here instead of below.

**Comment by tdreyno** [comment]
So you're saying you want `\---` to be a valid delimiter?

**Comment by Arcovion** [comment]
>shows up as errors in editors such as Sublime Text and RubyMine

Ah, it's breaking the syntax highlighting/linting in your editor? So this is not actually an error in Middleman;  the test is just there to show that `\---` is now allowed as a delimiter?

In that case, [please fix the test](https://travis-ci.org/middleman/middleman/jobs/85046858#L3546) as mentioned above (I think you probably just need to change `When I go to "/front-matter.haml"` to `When I go to "/front-matter.html"`) and remove the `content.sub!(/\\---/, '---')` line as it's not needed.

Will let @tdreyno decide if this should be added.

File: middleman-core/lib/middleman-core/util/data.rb (hunk 2 above)
Comment (Arcovion) [inline middleman-core/lib/middleman-core/util/data.rb]: Not necessary, the `frontmatter` variable already has the delimiters stripped away.

**Comment by DarylWright** [comment]
Exactly, I do want \--- to be a valid delimiter since that's the only way that the hyphen character can get escaped in Haml if it's the first character in a line. Middleman still works functionally when --- is used in Haml files, but that is not a correct method of adding frontmatter since it leaves the file in an errored state based on the Haml spec.

File: middleman-core/lib/middleman-core/util/data.rb (hunk 2 above)
Comment (DarylWright) [inline middleman-core/lib/middleman-core/util/data.rb]: You're right, thanks

**Comment by Arcovion** [comment]
Thing is, there's a world of filetypes out there that would need different delimiters and stuff, not just Haml. The yaml data itself should be a syntax error too, not just the delimiters.

My editor highlights frontmatter correctly for markdown and some other filetypes - this is an editor problem, nothing to do with the Haml spec.

**Comment by DarylWright** [comment]
> Thing is, there's a world of filetypes out there that would need different delimiters and stuff, not just Haml. 

While true, there's a limited number of file types supported in Middleman, and as far as I know, Haml is the only one that needs this particular delimiter.

> The yaml data itself should be a syntax error too, not just the delimiters.

Not really, the yaml data is just treated as plain text in Haml.

> My editor highlights frontmatter correctly for markdown and some other filetypes - this is an editor problem, nothing to do with the Haml spec.

Again, Haml is the only file type supported by Middleman that breaks a syntax rule by including yaml frontmatter. Consider for a moment that this is actually an editor problem. What does it mean for us to ask an editor to ignore a syntax rule for a particular file type? How then would we guarantee that files made with such an editor will work for someone else? What if we want to use the Haml files elsewhere without a bunch of hackery? What do you think?

**Comment by DarylWright** [comment]
I think I have the frontmatter fixture set up correctly as of ea67ec130d30c8ad18b08339f82aef8e23df786e, but I can't for the life of me figure out why the test still fails. Seems the Haml template isn't getting converted to HTML and therefore isn't found. Am I missing something?

```
Scenario: Rendering Haml (yaml)                    # features/front-matter.feature:31
    Given the Server is running at "frontmatter-app" # lib/middleman-core/step_definitions/server_steps.rb:57
    When I go to "/front-matter-haml.html"           # lib/middleman-core/step_definitions/server_steps.rb:66
    Then I should see "<h1>This is the title</h1>"   # lib/middleman-core/step_definitions/server_steps.rb:78
      expected "<html><head></head><body><h1>File Not Found</h1><p>/front-matter-haml.html</p></body></html>" to include "<h1>This is the title</h1>" (RSpec::Expectations::ExpectationNotMetError)
      ./lib/middleman-core/step_definitions/server_steps.rb:79:in `/^I should see "([^\"]*)"$/'
      features/front-matter.feature:34:in `Then I should see "<h1>This is the title</h1>"'
    Then I should not see "---"                      # lib/middleman-core/step_definitions/server_steps.rb:90
```

**Comment by Arcovion** [comment]
I think the file needs to be renamed: `front-matter-haml.haml` &rarr; `front-matter-haml.html.haml`

**Comment by DarylWright** [comment]
Thanks @Arcovion, that allowed the HTML to be output at least. The test still fails because the frontmatter hasn't been parsed out. My regex changes will likely have to be corrected.

**Comment by DarylWright** [comment]
The last commit passes my test, are we good to go?

**Comment by tdreyno** [comment]
@Arcovion is there any way this could become a config option so anyone can pick different delimiters and all the variations don't have to be in core? Or is the regex too complicated?

**Comment by Arcovion** [comment]
Sure, that'd be easy. What would it look like in the config?
```ruby
config[:frontmatter_delims] = {json: [%w[;;; ;;;]], yaml: [%w[--- ---], %w[--- ...]]}
config[:frontmatter_delims][:yaml] << %w[\\--- \\---]
```
Hard part for me would be testing it, and idk off by heart how to get config options from within the `Middleman::Util::Data::parse` method.

**Comment by tdreyno** [comment]
`app.config[:frontmatter_delims][:yaml]` should do it. Seems like the tests provided in this PR might handle the rest.

**Comment by DarylWright** [comment]
I guess that's a successful first pull request then?

**Comment by Arcovion** [comment]
@tdreyno That raises:
```ruby
undefined local variable or method `app' for Middleman::Util::Data:Module (NameError)
```
This works:
```ruby
Middleman::Application.config[:frontmatter_delims]
```
But only picks up the ones from `application.rb` in `define_setting`:
```ruby
# Set textual delimiters that denote the start and end of frontmatter
define_setting :frontmatter_delims, {
  json: [%w(;;; ;;;)],
  yaml: [%w(--- ---), %w(--- ...)]
}, 'Allowed frontmatter delimiters'
```
Not the stuff set in `middleman/middleman-core/fixtures/frontmatter-app/config.rb`,
which is where I'm doing:
```ruby
config[:frontmatter_delims][:yaml] << %w(\\--- \\---)
```
Setting it to *anything* there has no effect.
Is it even possible to get config options in that method after configuration?

**Comment by tdreyno** [comment]
@Arcovion I'll poke around

**Comment by tdreyno** [comment]
@Arcovion The non-instanced path (`Application`) is the default settings, they are immutable.

Regarding `app`, I thought you were in the front matter extension, not the util file.

You'll just have to add another parameter and pass it from further up where needed.

**Comment by DarylWright** [comment]
Thanks for considering these changes!

**Pull request closed without merging**